### ISSUE TITLE
Discontinuing Python 3.5 support and adding Python 3.7 support

### ISSUE THREAD

**Issue by jchodera** [openmm/openmm#2180]
Heads-up:
https://twitter.com/condaforge/status/1045699331171528707?s=19

> conda-forge is planning on stopping new Python 3.5 builds by default after Oct. 1, 2018.  Please let us know now if this is problematic for you! This is consistent with our policy of building two versions of Python 3 (v3.6 and v3.7) and Python 2.7

We should do the same after OpenMM 7.3 to stay consistent with conda-forge or else we will have a huge maintenance difficulty on our hands.

**Comment by peastman** [comment]
In combination with this, we also should add Python 3.7 packages.  In fact, it would be good to do that even for 7.3.

The other thing we should investigate is whether we can use the stable ABI (https://docs.python.org/3/c-api/stable.html) to create a single package that works with any Python 3.x.

**Comment by jchodera** [comment]
Working on py37 builds for the `dev` package in [this PR](https://github.com/omnia-md/conda-dev-recipes/pull/153). If this works, we can build 7.3 with it too.

I haven't seen guidelines for building universal `conda` Python packages by indicating that the stable ABI subset is being used. Maybe we can investigate after 7.3? I imagine we might have to completely deprecate Python 2.x if we want to build a package that doesn't have a Python version dependency, so we would want to warn people that 7.3 is the last package that supports 2.x first.

**Comment by jchodera** [comment]
I'm still working on py37 builds but conda-forge packages like [this one](https://github.com/conda-forge/sphinxcontrib-autodoc_doxygen-feedstock/issues/3) have still not been re-rendered for Python 3.7. I'm not sure what conda-forge's plan to rebuild packages for Python 3.7 actually is, or even where they would discuss/announce such a plan.

**Comment by jchodera** [comment]
@peastman: I'm making progress with the python 3.7 build, but it looks like we need to get the OpenMM recipe for building the docs to work with the latest version of [sphinx on anaconda conda-forge](https://anaconda.org/conda-forge/sphinx/files) to get this to work. Can you see if you can debug what might be going on when you install the latest sphinx?

This is the error trace we end up with using the latest sphinx:
```
+ make -j2 sphinxpdf
Scanning dependencies of target sphinxpdf
[  0%] Generating PDF user guide
[100%] Generating PDF developer guide
gmake[4]: warning: jobserver unavailable: using -j1.  Add `+' to parent make rule.
gmake[4]: warning: jobserver unavailable: using -j1.  Add `+' to parent make rule.
Traceback (most recent call last):
  File "/opt/conda/conda-bld/openmm_1539369821245/_b_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/bin/sphinx-build", line 11, in <module>
    sys.exit(main())
  File "/opt/conda/conda-bld/openmm_1539369821245/_b_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/lib/python3.7/site-packages/sphinx/cmd/build.py", line 313, in main
    locale.setlocale(locale.LC_ALL, '')
  File "/opt/conda/conda-bld/openmm_1539369821245/_b_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/lib/python3.7/locale.py", line 604, in setlocale
    return _setlocale(category, locale)
locale.Error: unsupported locale setting
Traceback (most recent call last):
  File "/opt/conda/conda-bld/openmm_1539369821245/_b_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/bin/sphinx-build", line 11, in <module>
    sys.exit(main())
  File "/opt/conda/conda-bld/openmm_1539369821245/_b_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/lib/python3.7/site-packages/sphinx/cmd/build.py", line 313, in main
    locale.setlocale(locale.LC_ALL, '')
  File "/opt/conda/conda-bld/openmm_1539369821245/_b_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/lib/python3.7/locale.py", line 604, in setlocale
    return _setlocale(category, locale)
locale.Error: unsupported locale setting
gmake[4]: *** [latexpdf] Error 1
make[3]: *** [sphinx-docs/userguide/latex/OpenMMUsersGuide.pdf] Error 2
make[3]: *** Waiting for unfinished jobs....
gmake[4]: *** [latexpdf] Error 1
make[3]: *** [sphinx-docs/developerguide/latex/OpenMMDeveloperGuide.pdf] Error 2
make[2]: *** [docs-source/CMakeFiles/sphinxpdf.dir/all] Error 2
make[1]: *** [docs-source/CMakeFiles/sphinxpdf.dir/rule] Error 2
make: *** [sphinxpdf] Error 2
Traceback (most recent call last):
  File "/opt/conda/bin/conda-build", line 6, in <module>
    sys.exit(conda_build.cli.main_build.main())
  File "/opt/conda/lib/python3.6/site-packages/conda_build/cli/main_build.py", line 342, in main
    execute(sys.argv[1:])
  File "/opt/conda/lib/python3.6/site-packages/conda_build/cli/main_build.py", line 333, in execute
    noverify=args.no_verify)
  File "/opt/conda/lib/python3.6/site-packages/conda_build/api.py", line 97, in build
    need_source_download=need_source_download, config=config)
  File "/opt/conda/lib/python3.6/site-packages/conda_build/build.py", line 1524, in build_tree
    config=config)
  File "/opt/conda/lib/python3.6/site-packages/conda_build/build.py", line 1147, in build
    utils.check_call_env(cmd, env=env, cwd=src_dir)
  File "/opt/conda/lib/python3.6/site-packages/conda_build/utils.py", line 628, in check_call_env
    return _func_defaulting_env_to_os_environ(subprocess.check_call, *popenargs, **kwargs)
  File "/opt/conda/lib/python3.6/site-packages/conda_build/utils.py", line 624, in _func_defaulting_env_to_os_environ
    return func(_args, **kwargs)
  File "/opt/conda/lib/python3.6/subprocess.py", line 291, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/bin/bash', '-x', '-e', '/opt/conda/conda-bld/openmm_1539369821245/work/conda_build.sh']' returned non-zero exit status 2.
```

**Comment by peastman** [comment]
I just updated Sphinx on my computer to 1.8.1 and ran `make sphinxpdf`.  It worked without any problems.

**Comment by peastman** [comment]
Here are a couple of pages where people are reporting the same error message in different situations.  One of them might help.

https://stackoverflow.com/questions/14547631/python-locale-error-unsupported-locale-setting/36257050
https://stackoverflow.com/questions/36394101/pip-install-locale-error-unsupported-locale-setting

**Comment by jchodera** [comment]
@peastman: I think we figured out the Python 3.7 issues, thanks to @lnaden's help.

I've just merged the changes in with https://github.com/omnia-md/conda-dev-recipes/pull/155 and am building a nightly `master` dev build as 7.4.0. It should post to anaconda cloud within the hour.

https://anaconda.org/omnia/openmm/files

Presuming this works, I should be able to upgrade the `conda-recipes` build system and then merge https://github.com/omnia-md/conda-recipes/pull/911 so we can get a release candidate built!

**Comment by jchodera** [comment]
Python 3.7 `7.4.0` builds are posted to the `devcudaXY` label! Please test!

**Comment by swails** [comment]
I'm currently seeing problems with Python 3.7 in #2198 -- see the error message [here](https://jenkins.jasonswails.com/blue/organizations/jenkins/GH-pandegroup%2FOpenMM/detail/OpenMM/37/pipeline#step-50-log-1017).

Basically, the PDBx parsers supplied by RCSB relies on a generator raising `StopIteration`, something that Python deprecated in Python 3.5 and turned into a hard error in Python 3.7.  See PEP 479 for more details.

Basically what used to be a StopIteration is now a RuntimeError.  I just fixed this in #2198.

**Comment by peastman** [comment]
Thanks!  Should we try to fix it separately, or just wait for #2198?  I see you're aiming for a record on "most commits in a single PR". 😀

**Comment by swails** [comment]
> I see you're aiming for a record on "most commits in a single PR". 😀

I'm aiming to smash that record 🔨 

> Should we try to fix it separately, or just wait for #2198?

I'd elect to just wait.  I think I'm pretty close.  Of course if a release is imminent and someone is blocked by this particular issue (unlikely given that it only affects PDBx) we can raise it separately.

I really want to get the whole build running in a docker agent, since that will make it quite a bit more portable (not to mention that it supplies a Dockerfile that creates an OpenMM-ready container).

**Comment by jchodera** [comment]
We've already released 7.3.0, so we can issue 7.3.1 when we get the fix merged and tested.

**Comment by jchodera** [comment]
@peastman: Do you happen to know if we fixed this in 7.3.1?

**Comment by peastman** [comment]
No, it didn't. :(  The PR didn't mention anything about bug fixes, just changes to jenkins, so it didn't get included in 7.3.1.
